examples/cpp/src/cagra_streaming_example.cu

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2024-2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cstdint>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+
+#include <thrust/sequence.h>
+
+#include <cuvs/neighbors/cagra.hpp>
+#include <cuvs/neighbors/cagra_optimize.hpp>
+#include <cuvs/neighbors/ivf_pq.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/device_resources.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+/*
+Streaming CAGRA index build example
+-----------------------------------
+
+This example shows how to build a CAGRA graph index by streaming host batches:
+1) Stage the first few batches to train an IVF-PQ index.
+2) Incrementally extend the IVF-PQ index with every batch.
+3) Run IVF-PQ search over the full dataset to form an intermediate k-NN graph.
+4) Optimize that graph into the final fixed-degree CAGRA graph.
+
+The dataset is kept on the host the whole time to mimic a scenario where the
+full corpus cannot live on the device.
+*/
+
+namespace {
+
+void make_host_dataset(raft::host_matrix_view<float, int64_t, raft::row_major> dataset)
```

Comment:
Can we maybe call this "generate host dataset"? Honestly, I think I'd prefer to use `make_blobs` from `raft` than to just generate completely random uniform vectors. make_blobs at least has some locality to the vector space that can be exploited by IVFPQ and CAGRA.